design/0005-protocol-detection.md

```diff
@@ -0,0 +1,308 @@
+
+The destination controller uses this annotation when assisting the outbound
+proxy's protocol discovery via the new `Destination.Strategy` API endpoint
+(to be added as a part of TCP work), this annotation instructs the proxy to
+skip protocol detection on this stream and to forward it opaquely. This
+should be honored for both meshed and un-meshed workloads.
+
+When resolving a `Service` address, the union of all values from all pods in
+that service is considered.
+
+When communicating with a meshed endpoint, the destination response includes
+a protocol upgrade hint including the remote inbound proxy's port number and
+its identity (for use with mTLS).
+
+The outbound proxy then connects to the remote inbound port _directly_ (and
+not the original destination port). The endpoint's application port (as
+provided by the discovery response and **not** the `SO_ORIG_DST` lookup),
+must be preserved and transmitted to the remote proxy so that it can route it
+properly. This is transmitted via a new _Linkerd Connection Header_.
+
+#### Multi-Cluster Service Mirror Controller
+
+The service-mirror creates local endpoints that reflect the opaqueness, as
+configured in the target cluster. The multi-cluster gateway then uses the
+connection header to route the connection through the gateway.
+
+### Proxy handling of inbound-targeted connections
+
+The proxy makes assumptions about traffic that targets its inbound proxy
+server (as opposed to traffic that targets an application port, but is
+rerouted through the inbound proxy port). A connection that targets the
+inbound proxy port:
+
+* Is assumed to originate from other proxy processes;
+* MUST use Linkerd's mTLS identity; and
+* SHOULD include a _Linkerd Connection Header_ preamble.
+
+Note that outbound proxies _must_ set a connection preamble when targeting an
+inbound proxy; but, in order for Linkerd multi-cluster gateways to remain
+backwards-compatible, the inbound proxy cannot assume the presence of this
+preamble and should revert to protocol detection in such a case.
+
+#### Linkerd Connection Header preamble
+
+The connection header is a preamble that is transmitted from outbound proxies
+to remote inbound proxies. This preamble consists of:
+
+* a 12B magic marker;
+* a 4B message length; and
+* a variable-length protobuf-encoded message.
+
+When the first 16B of a stream match the magic marker, the proxy strips the
+entire header from the connection before proxying it to the application.
+
+These messaged are prefaced with a magic marker prefix: `l5d.io/proxy`, or
+`0x6c35642e696f2f70726f7879`.
+
+Following that, a 4-byte unsigned integer (Big Endian) message length,
+followed by a protobuf-encoded `LinkerdConnectionHeaderV1` message.
+
+This message type includes at least an optional application port, and an
+optional detection hint.
+
+When the port is set, it identifies a port on localhost to which the traffic
+should be forwarded. When unset (`0`) or set to the value of the inbound or
```

Comment:
If we are using a proto structure, I assume this can actually be expressed as an Optional val (I think), rather thansetting it to 0. WDYT ?

Reply:
When using a simple value like
```proto3
int port = 1;
```
the default value is 0.

If you wanted to have it manifest as an option, you have to do something like:
```proto3
message Port {
  int port = 1;
}
Port port = 1;
```

I don't think there's any real value in the compexity.  We have to handle a 0 value either way.